runtime/src/lib.rs

```diff
@@ -134,73 +134,72 @@ pub unsafe extern "C" fn cairo_native__libfunc__hades_permutation(
 }
 
 /// Felt252 type used in cairo native runtime
-pub type FeltDict = (HashMap<[u8; 32], NonNull<std::ffi::c_void>>, u64);
+#[derive(Debug, Default)]
+pub struct FeltDict {
+    pub inner: HashMap<[u8; 32], NonNull<std::ffi::c_void>>,
+    pub count: u64,
+}
 
-/// Allocates a new dictionary. Internally a rust hashmap: `HashMap<[u8; 32], NonNull<()>`
+/// Allocate a new dictionary.
 ///
 /// # Safety
 ///
 /// This function is intended to be called from MLIR, deals with pointers, and is therefore
 /// definitely unsafe to use manually.
 #[no_mangle]
-pub unsafe extern "C" fn cairo_native__alloc_dict() -> *mut std::ffi::c_void {
-    Box::into_raw(Box::<FeltDict>::default()) as _
+pub unsafe extern "C" fn cairo_native__dict_new() -> *mut FeltDict {
+    Box::into_raw(Box::<FeltDict>::default())
 }
 
-/// Frees the dictionary.
+/// Free a dictionary using an optional callback to drop each element.
 ///
 /// # Safety
 ///
 /// This function is intended to be called from MLIR, deals with pointers, and is therefore
 /// definitely unsafe to use manually.
 #[no_mangle]
-pub unsafe extern "C" fn cairo_native__dict_free(ptr: *mut FeltDict) {
-    let mut map = Box::from_raw(ptr);
+pub unsafe extern "C" fn cairo_native__dict_drop(
+    ptr: *mut FeltDict,
+    drop_fn: Option<extern "C" fn(*mut std::ffi::c_void)>,
```

Comment:
I would add a comment here, clarifying that this Option is ffi safe due to this: https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization

Reply:
Done